File: authfn/core/src/core/api-keys.ts
import { randomBytes } from 'node:crypto';
import type { AuthFnApiKeyRecord, AuthFnConfig, AuthFnSession } from '../types.js';
import { AuthFnApiKeyRevokedError, AuthFnNotFoundError, AuthFnValidationError } from './errors.js';
import { hashSecret } from './sessions.js';

export interface CreateApiKeyInput {
  userId: string;
  name: string;
  scopes?: string[];
  metadata?: Record<string, unknown>;
  expiresAt?: Date;
}

export interface CreatedApiKey {
  keyId: string;
  secret: string;
  record: AuthFnApiKeyRecord;
}

export interface SanitizedApiKey {
  id: string;
  userId?: string | null;
  name?: string;
  scopes?: string[];
  metadata?: Record<string, unknown>;
  expiresAt?: Date | null;
  revokedAt?: Date | null;
  lastUsedAt?: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

export async function createApiKey(
  config: Pick<AuthFnConfig, 'database' | 'namespace'>,
  input: CreateApiKeyInput,
  options?: { now?: () => Date; secretPrefix?: string }
): Promise<CreatedApiKey> {
  assertValidApiKeyName(input.name);
  assertValidApiKeyExpiry(input.expiresAt);
  const now = options?.now?.() ?? new Date();
  const secret = createApiKeySecret(options?.secretPrefix);
  const record: AuthFnApiKeyRecord = {
    id: createIdentifier('key'),
    userId: input.userId,
    name: input.name,
    secretHash: hashSecret(secret),
    scopes: input.scopes,
    metadata: input.metadata,
    expiresAt: input.expiresAt ?? null,
    revokedAt: null,
    lastUsedAt: null,
    createdAt: now,
    updatedAt: now
  };

  await config.database.create<AuthFnApiKeyRecord>({
    model: 'api_keys',
    data: record,
    namespace: namespace(config)
  });

  return {
    keyId: record.id,
    secret,
    record
  };
}

export async function listApiKeysForUser(
  config: Pick<AuthFnConfig, 'database' | 'namespace'>,
  userId: string
): Promise<SanitizedApiKey[]> {
  const records = await config.database.findMany<AuthFnApiKeyRecord>({
    model: 'api_keys',
    where: [{ field: 'userId', operator: 'eq', value: userId }],
    orderBy: [
      { field: 'createdAt', direction: 'asc' },
      { field: 'id', direction: 'asc' }
    ],
    namespace: namespace(config)
  });

  return records.map((record) => sanitizeApiKeyRecord(record));
}

export async function findApiKeyById(
  config: Pick<AuthFnConfig, 'database' | 'namespace'>,
  keyId: string
): Promise<AuthFnApiKeyRecord | null> {
  return config.database.findOne<AuthFnApiKeyRecord>({
    model: 'api_keys',
    where: [{ field: 'id', operator: 'eq', value: keyId }],
    namespace: namespace(config)
  });
}

export async function revokeApiKeyById(
  config: Pick<AuthFnConfig, 'database' | 'namespace'>,
  keyId: string,
  options?: { userId?: string; now?: () => Date }
): Promise<AuthFnApiKeyRecord> {
  const record = await findApiKeyById(config, keyId);
  if (!record || (options?.userId && record.userId !== options.userId)) {
    throw new AuthFnNotFoundError('API key not found', { keyId });
  }

  if (record.revokedAt) {
    return record;
  }

  const revokedAt = options?.now?.() ?? new Date();
  return config.database.update<AuthFnApiKeyRecord>({
    model: 'api_keys',
    where: [{ field: 'id', operator: 'eq', value: keyId }],
    data: {
      revokedAt,
      updatedAt: revokedAt
    },
    namespace: namespace(config)
  });
}

export async function authenticateApiKey(
  config: Pick<AuthFnConfig, 'database' | 'namespace'>,
  secret: string,
  options?: { now?: () => Date }
): Promise<AuthFnSession | null> {
  const record = await config.database.findOne<AuthFnApiKeyRecord>({
    model: 'api_keys',
    where: [{ field: 'secretHash', operator: 'eq', value: hashSecret(secret) }],
    namespace: namespace(config)
  });

  if (!record) {
    return null;
  }

  if (record.revokedAt) {
    throw new AuthFnApiKeyRevokedError('API key has been revoked', {
      keyId: record.id
    });
  }

  const now = options?.now?.() ?? new Date();
  if (record.expiresAt && record.expiresAt.getTime() <= now.getTime()) {
    return null;
  }

  const updatedRecord = await config.database.update<AuthFnApiKeyRecord>({
    model: 'api_keys',
    where: [{ field: 'id', operator: 'eq', value: record.id }],
    data: {
      lastUsedAt: now,
      updatedAt: now
    },
    namespace: namespace(config)
  });

  return {
    id: updatedRecord.id,
    type: 'api-key',
    subject: {
      actorId: updatedRecord.id,
      actorType: 'api-key'
    },
    actorType: 'api-key',
    actorId: updatedRecord.id,
    resourceIds: [],
    methods: ['api-key'],
    expiresAt: updatedRecord.expiresAt ?? undefined,
    metadata: {
      ...(updatedRecord.metadata ?? {}),
      ownerUserId: updatedRecord.userId ?? undefined,
      scopes: updatedRecord.scopes ?? []
    }
  };
}

export function sanitizeApiKeyRecord(record: AuthFnApiKeyRecord): SanitizedApiKey {
  return {
    id: record.id,
    userId: record.userId,
    name: record.name,
    scopes: record.scopes,
    metadata: record.metadata,
    expiresAt: record.expiresAt ?? null,
    revokedAt: record.revokedAt ?? null,
    lastUsedAt: record.lastUsedAt ?? null,
    createdAt: record.createdAt,
    updatedAt: record.updatedAt
  };
}

function createApiKeySecret(prefix: string = 'ak'): string {
  return `${prefix}_${randomBytes(24).toString('base64url')}`;
}

function createIdentifier(prefix: string): string {
  return `${prefix}_${randomBytes(8).toString('hex')}`;
}

function namespace(config: Pick<AuthFnConfig, 'namespace'>): string {
  return config.namespace ?? 'authfn';
}

function assertValidApiKeyName(name: string): void {
  if (typeof name !== 'string' || name.trim().length === 0) {
    throw new AuthFnValidationError('API key name is required', {
      field: 'name'
    });
  }

  const bytes = Buffer.byteLength(name, 'utf8');
  if (bytes > 128) {
    throw new AuthFnValidationError('API key name must be 128 UTF-8 bytes or fewer', {
      field: 'name',
      maxBytes: 128,
      receivedBytes: bytes
    });
  }
}

function assertValidApiKeyExpiry(expiresAt: Date | undefined): void {
  if (!expiresAt) {
    return;
  }

  if (!(expiresAt instanceof Date) || Number.isNaN(expiresAt.getTime())) {
    throw new AuthFnValidationError('API key expiry must be a valid date', {
      field: 'expiresAt'
    });
  }
}

File: authfn/python/authfn/plugins/api_keys.py
"""API key plugin and service for authfn Python."""

from __future__ import annotations

import hashlib
import secrets
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from ..types import (
    ApiKeyRevokedError,
    AuthFnConfig,
    AuthFnPlugin,
    AuthFnSession,
    ExpiredCredentialsError,
    ValidationError,
)


def _default_now() -> datetime:
    return datetime.now(timezone.utc)


def _create_id(prefix: str) -> str:
    return f"{prefix}_{secrets.token_hex(8)}"


def _hash_secret(secret: str) -> str:
    return hashlib.sha256(secret.encode("utf-8")).hexdigest()


def _as_utc(value: Optional[datetime]) -> Optional[datetime]:
    if value is None:
        return None
    if value.tzinfo is None:
        return value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc)


def _read_authorization_secret(request: Any) -> Optional[str]:
    headers = getattr(request, "headers", {}) or {}
    authorization = headers.get("authorization") or headers.get("Authorization")
    if not authorization:
        return None
    value = authorization.strip()
    if value.lower().startswith("bearer "):
        return value[7:].strip() or None
    if value.lower().startswith("api-key "):
        return value[8:].strip() or None
    return None


def _assert_valid_name(name: str) -> None:
    if not name:
        raise ValidationError("API key name is required")
    length = len(name.encode("utf-8"))
    if length > 128:
        raise ValidationError("API key name must be 128 UTF-8 bytes or fewer")


@dataclass
class ApiKeyPluginConfig:
    now: Any = _default_now
    secret_prefix: str = "ak"


class ApiKeyService:
    def __init__(self, config: AuthFnConfig, plugin_config: Optional[ApiKeyPluginConfig] = None):
        self.config = config
        self.plugin_config = plugin_config or ApiKeyPluginConfig()

    async def create_key(
        self,
        *,
        user_id: Optional[str] = None,
        name: str,
        scopes: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        expires_at: Optional[datetime] = None,
        resource_ids: Optional[List[str]] = None,
    ) -> Dict[str, Any]:
        _assert_valid_name(name)

        stored_metadata = dict(metadata or {})
        if resource_ids is not None:
            stored_metadata.setdefault("resourceIds", list(resource_ids))

        secret = f"{self.plugin_config.secret_prefix}_{secrets.token_urlsafe(24)}"
        now = self.plugin_config.now()
        normalized_expires_at = _as_utc(expires_at)
        record = {
            "id": _create_id("key"),
            "userId": user_id,
            "name": name,
            "secretHash": _hash_secret(secret),
            "scopes": list(scopes or []),
            "metadata": stored_metadata,
            "expiresAt": normalized_expires_at,
            "revokedAt": None,
            "lastUsedAt": None,
            "createdAt": now,
            "updatedAt": now,
        }
        await self.config.database.create(
            model="api_keys",
            data=record,
            namespace=self.config.namespace,
        )
        return {"keyId": record["id"], "secret": secret, "record": record}

    async def list_keys(self, *, user_id: str) -> List[Dict[str, Any]]:
        rows = await self.config.database.find_many(
            model="api_keys",
            where=[{"field": "userId", "operator": "eq", "value": user_id}],
            order_by=[{"field": "createdAt", "direction": "asc"}],
            namespace=self.config.namespace,
        )
        return [self._sanitize(row) for row in rows]

    async def revoke_key(self, *, key_id: str, user_id: Optional[str] = None) -> Dict[str, Any]:
        key = await self.config.database.find_one(
            model="api_keys",
            where=[{"field": "id", "operator": "eq", "value": key_id}],
            namespace=self.config.namespace,
        )
        if key is None or (user_id is not None and key.get("userId") != user_id):
            raise ValidationError("API key not found")
        if key.get("revokedAt") is not None:
            return key
        now = self.plugin_config.now()
        return await self.config.database.update(
            model="api_keys",
            where=[{"field": "id", "operator": "eq", "value": key_id}],
            data={"revokedAt": now, "updatedAt": now},
            namespace=self.config.namespace,
        )

    async def authenticate(self, request: Any) -> Optional[AuthFnSession]:
        secret = _read_authorization_secret(request)
        if secret is None:
            return None
        row = await self.config.database.find_one(
            model="api_keys",
            where=[{"field": "secretHash", "operator": "eq", "value": _hash_secret(secret)}],
            namespace=self.config.namespace,
        )
        if row is None:
            return None
        if row.get("revokedAt") is not None:
            raise ApiKeyRevokedError("API key has been revoked")
        expires_at = _as_utc(row.get("expiresAt"))
        now = self.plugin_config.now()
        if now.tzinfo is None:
            now = now.replace(tzinfo=timezone.utc)
        else:
            now = now.astimezone(timezone.utc)
        if expires_at is not None and expires_at <= now:
            raise ExpiredCredentialsError("API key has expired")
        await self.config.database.update(
            model="api_keys",
            where=[{"field": "id", "operator": "eq", "value": row["id"]}],
            data={"lastUsedAt": now, "updatedAt": now},
            namespace=self.config.namespace,
        )
        metadata = dict(row.get("metadata") or {})
        return AuthFnSession.model_validate(
            {
                "id": row["id"],
                "type": "api-key",
                "actorType": "api-key",
                "actorId": row["id"],
                "resourceIds": metadata.get("resourceIds"),
                "methods": ["api-key"],
                "expiresAt": expires_at,
                "metadata": {
                    "ownerUserId": row.get("userId"),
                    "scopes": row.get("scopes") or [],
                    **metadata,
                },
            }
        )

    def _sanitize(self, row: Dict[str, Any]) -> Dict[str, Any]:
        return {
            "id": row["id"],
            "userId": row.get("userId"),
            "name": row.get("name"),
            "scopes": row.get("scopes"),
            "metadata": row.get("metadata"),
            "expiresAt": row.get("expiresAt"),
            "revokedAt": row.get("revokedAt"),
            "lastUsedAt": row.get("lastUsedAt"),
            "createdAt": row.get("createdAt"),
            "updatedAt": row.get("updatedAt"),
        }


def authfn_api_key_plugin(config: Optional[ApiKeyPluginConfig] = None) -> AuthFnPlugin:
    resolved = config or ApiKeyPluginConfig()
    plugin = AuthFnPlugin(
        name="apiKey",
        schema_factory=lambda _cfg: [
            {
                "modelName": "api_keys",
                "fields": {
                    "id": {"type": "string", "required": True, "fieldName": "id"},
                    "userId": {"type": "string", "required": False, "fieldName": "user_id"},
                    "name": {"type": "string", "required": False, "fieldName": "name"},
                    "secretHash": {
                        "type": "string",
                        "required": True,
                        "fieldName": "secret_hash",
                    },
                    "scopes": {"type": "json", "required": False, "fieldName": "scopes"},
                    "metadata": {"type": "json", "required": False, "fieldName": "metadata"},
                    "expiresAt": {"type": "date", "required": False, "fieldName": "expires_at"},
                    "revokedAt": {"type": "date", "required": False, "fieldName": "revoked_at"},
                    "lastUsedAt": {"type": "date", "required": False, "fieldName": "last_used_at"},
                    "createdAt": {"type": "date", "required": True, "fieldName": "created_at"},
                    "updatedAt": {"type": "date", "required": True, "fieldName": "updated_at"},
                },
                "indexes": [
                    {"name": "idx_authfn_api_keys_secret_hash", "fields": ["secretHash"]},
                    {
                        "name": "idx_authfn_api_keys_user_id_created_at",
                        "fields": ["userId", "createdAt"],
                    },
                ],
            }
        ],
        routes_factory=lambda _ctx: [
            {"method": "POST", "path": "/api-keys"},
            {"method": "GET", "path": "/api-keys"},
            {"method": "DELETE", "path": "/api-keys/:keyId"},
        ],
    )
    plugin._authfn_config = resolved
    return plugin


__all__ = [
    "ApiKeyPluginConfig",
    "ApiKeyService",
    "authfn_api_key_plugin",
]
